cmd/mt-whisper-importer-reader/main.go

```diff
@@ -0,0 +1,339 @@
+			continue
+		}
+
+		b, err := met.MarshalCompressed()
+		if err != nil {
+			throwError(fmt.Sprintf("Failed to encode metric: %q", err))
+			continue
+		}
+
+		req, err := http.NewRequest("POST", *httpEndpoint, io.Reader(b))
+		if err != nil {
+			panic(fmt.Sprintf("Cannot construct request to http endpoint %q: %q", *httpEndpoint, err))
+		}
+		req.Header.Set("Content-Type", "application/json")
+		req.Header.Set("Content-Encoding", "gzip")
+
+		_, err = client.Do(req)
+		if err != nil {
+			throwError(fmt.Sprintf("Error sending request to http endpoint %q: %q", *httpEndpoint, err))
+			continue
+		}
+	}
+	wg.Done()
+}
+
+// generate the metric name based on the file name and given prefix
+func getMetricName(file string) string {
+	// remove all leading '/' from file name
+	for file[0] == '/' {
+		file = file[1:]
+	}
+
+	return *namePrefix + strings.Replace(strings.TrimSuffix(file, ".wsp"), "/", ".", -1)
+}
+
+// pointSorter sorts points by timestamp
+type pointSorter []whisper.Point
+
+func (a pointSorter) Len() int           { return len(a) }
+func (a pointSorter) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a pointSorter) Less(i, j int) bool { return a[i].Timestamp < a[j].Timestamp }
+
+// the whisper archives are organized like a ringbuffer. since we need to
+// insert the points into the chunks in order we first need to sort them
+func sortPoints(points pointSorter) pointSorter {
+	sort.Sort(points)
+	return points
+}
+
+func getMetric(w *whisper.Whisper, file string) (*archive.Metric, error) {
+	if len(w.Header.Archives) == 0 {
+		return nil, errors.New(fmt.Sprintf("ERROR: Whisper file contains no archives: %q", file))
+	}
+
+	archives := make([]archive.Archive, 0, len(w.Header.Archives))
+	var chunkSpan uint32
+	var rowKey string
+	name := getMetricName(file)
+
+	// md gets generated from the first archive in the whisper file
+	md := getMetricData(name, int(w.Header.Archives[0].SecondsPerPoint))
+
+	for archiveIdx, archiveInfo := range w.Header.Archives {
+		if archiveIdx > 0 {
+			rowKey = api.AggMetricKey(
+				md.Id,
+				whisper.AggregationMethod(w.Header.Metadata.AggregationMethod).String(),
+				archiveInfo.SecondsPerPoint,
+			)
+		} else {
+			rowKey = md.Id
+		}
+
+		// only read archive if archiveIdx is in readArchives
+		if _, ok := readArchives[archiveIdx]; !ok && len(readArchives) > 0 {
+			continue
+		}
+
+		encodedChunks := make([]chunk.IterGen, 0)
+
+		if len(chunkSpans)-1 < archiveIdx {
+			// if we have more archives than chunk spans are specified, we simply use the last one
+			chunkSpan = chunkSpans[len(chunkSpans)-1]
+		} else {
+			chunkSpan = chunkSpans[archiveIdx]
+		}
+
+		points, err := w.DumpArchive(archiveIdx)
+		if err != nil {
+			return nil, errors.New(fmt.Sprintf("ERROR: Failed to read archive %d in %q, skipping: %q", archiveIdx, file, err))
+		}
+
+		var point whisper.Point
+		var t0, prevT0 uint32
+		var c *chunk.Chunk
+
+		for _, point = range sortPoints(points) {
+			if point.Timestamp == 0 {
```

Comment:
It shouldn't happen, but if it would happen then MT wouldn't handle it well: https://github.com/raintank/metrictank/blob/master/mdata/aggmetric.go#L378